composer/datasets/streaming/dataset.py

```diff
@@ -0,0 +1,379 @@
+
+    def _unpack_sample(self, data: bytes) -> Dict[str, Any]:
+        """Unpack a sample dict from raw bytes.
+
+        First unpacks the str to raw bytes dict, then unpacks each field's raw bytes.
+
+        Args:
+            data (bytes): The packed bytes of the sample.
+
+        Returns:
+            Dict[str, Any]: The sample dict.
+        """
+        key_to_raw = bytes_to_sample_dict(data, self.index.fields)
+        obj = {}
+        for key, decode in self.decoders.items():
+            value = key_to_raw[key]
+            if decode:
+                value = decode(value)
+            obj[key] = value
+        return obj
+
+    def __getitem__(self, idx: int) -> Any:
+        """Get the sample at the index, assuming its shard is loaded.
+
+        Do not call this directly unless all shards have been loaded. Will crash
+        if the shard is not loaded.
+
+        Args:
+            idx (int): Sample ID.
+
+        Returns:
+            Any: The sample.
+        """
+        shard = self.index.sample_shards[idx]
+        offset = self.index.sample_shard_offsets[idx]
+        size = self.index.bytes_per_sample[idx]
+
+        basename = get_shard_basename(shard)
+        shard_filename = os.path.join(self.local, basename)
+        fp = open(shard_filename, 'rb', 0)
+        fp.seek(offset)
+        data = fp.read(size)
+        fp.close()
+
+        return self._unpack_sample(data)
+
+    def _new_growing_epoch(self) -> int:
+        """Start a new growing epoch, in which we own the sample sequence because it grows.
+
+        Returns:
+            int: The epoch ID, an identifier which is given back to the caller.
+        """
+        with self._lock:
+            epoch = self._next_epoch
+            self._next_epoch += 1
+            self._epoch_to_todo_ids[epoch] = list(self._downloaded_ids)
+        return epoch
+
+    def _next_id(self, epoch: int) -> Optional[int]:
+        """Get next sample of the growing epoch given by epoch, or None if done.
+
+        If we are currently out of samples but not finished downloading the
+        shards, blocks until it has new samples.
+
+        Args:
+            epoch (int): The epoch, an identifier for this sequence of samples.
+
+        Returns:
+            int: ID of next sample.
+        """
+        while True:
+            with self._lock:
+                todo_ids = self._epoch_to_todo_ids[epoch]
+                if todo_ids:
+                    return todo_ids.pop(0)
+                elif self._are_all_shards_downloaded:
+                    del self._epoch_to_todo_ids[epoch]
+                    return None
+                else:
+                    pass
+            sleep(0.25)
+
+    def _iter_ids(self) -> Iterator[int]:
+        """Get an iterator over all our sample IDs.
+
+        Returns:
+            Iterator[int]: Each sample ID.
+        """
+        if self._lock is None:
+            self._lock = Lock()
```

Comment:
Can you add a comment explaining why the lock is created here rather than on `__init__` (I presume this is to be safe with multiple dataloader workers?)